allo/backend/aie.py

```diff
@@ -643,37 +786,59 @@ def codegen_aie_mlir(
-            else:
-                out_size_n_stride = f"[1, 1, {orig_out_shape[0]}, {orig_out_shape[1]}][0, 0, {orig_out_shape[1]}, 1]"
-            code += format_str(
-                f"aiex.npu.dma_memcpy_nd(0, 0, %arg{out_id}[0, 0, 0, 0]{out_size_n_stride}) {{id = 0 : i64, metadata = @out_sh_{kernel_name}}} : {orig_out_type}"
-            )
-            for arg_id in range(len(input_args) - 1):
+                arg_index += 1
+            for i, (_, orig_out_type, _, orig_out_shape) in enumerate(outputs):
+                arg_id = len(inputs) + i
+                if len(orig_out_shape) == 1:
+                    out_size_n_stride = f"[1, 1, 1, {orig_out_shape[0]}][0, 0, 0, 1]"
+                elif (
+                    len(mapping) == 2
+                    and len(orig_out_shape) == 2
+                    and dist_allocs[arg_id]
+                ):
+                    # now only support 2D mapping and 2D tensor
+                    out_size_n_stride = f"[{mapping[0]}, {mapping[1]}, {orig_out_shape[0] // mapping[0]}, {orig_out_shape[1] // mapping[1]}][{orig_out_shape[0] // mapping[0] * orig_out_shape[1]}, {orig_out_shape[1] // mapping[1]}, {orig_out_shape[1]}, 1]"
+                else:
+                    out_size_n_stride = f"[1, 1, {orig_out_shape[0]}, {orig_out_shape[1]}][0, 0, {orig_out_shape[1]}, 1]"
+                code += format_str(
+                    f"aiex.npu.dma_memcpy_nd(0, 0, %arg{arg_index}[0, 0, 0, 0]{out_size_n_stride}) {{id = 0 : i64, metadata = @out_sh_{kernel_name}}} : {orig_out_type}"
+                )
+                arg_index += 1
+            for arg_id in range(len(inputs)):
                 code += format_str(
-                    f"aiex.npu.dma_wait {{symbol = @in_sh_{kernel_name}_{arg_id}}}"
+                    f"aiex.npu.dma_wait {{symbol = @in_sh_{kernel_name}_a{arg_id}}}"
+                )
+            for i in range(len(outputs)):
+                arg_id = len(inputs) + i
+                code += format_str(
+                    f"aiex.npu.dma_wait {{symbol = @out_sh_{kernel_name}}}"
                 )
-            code += format_str(f"aiex.npu.dma_wait {{symbol = @out_sh_{kernel_name}}}")
     code += format_str("}")
     code += format_str("}", indent=2)
     code += "}"
     return code
 
 
-def get_kernel_index_ranges_from_mappings(kernel_mappings):
+def get_kernel_index_ranges(mod, kernel_mappings):
+    kernel_names = list(kernel_mappings.keys())
     kernel_index_ranges = {}
-    index = 0
-    for func_name, mapping in kernel_mappings.items():
-        start = index
-        end = index = start + math.prod(mapping)
-        kernel_index_ranges[func_name] = (start, end)
+    start = end = 0
+    k_id = 0
+    for func in list(mod.body.operations)[:-1]:
```

Comment:
Avoid using `list(mod.body.operations)[:-1]`. Use the helper function to get the core kernels